### PR TITLE
Migrate FSTBlobValue to FSTDelegateValue

### PR DESCRIPTION
**Pull request by wilhuff**
... and a bunch of related changes to use ByteString throughout.

Note the DIFFBASE on #3119.

### CODE REVIEW

File: Firestore/core/src/firebase/firestore/nanopb/byte_string.cc [hunk 1]
```diff
@@ -61,6 +61,10 @@ ByteString::ByteString(const std::vector<uint8_t>& value)
     : bytes_(MakeBytesArray(value.data(), value.size())) {
 }
 
+ByteString::ByteString(const pb_bytes_array_t* bytes)
+    : bytes_{MakeBytesArray(bytes->bytes, bytes->size)} {
```
Comment (rsgowman) [inline Firestore/core/src/firebase/firestore/nanopb/byte_string.cc]: HARD_ASSERT(bytes != nullptr)?

File: Firestore/core/src/firebase/firestore/remote/serializer.cc [hunk 2]
```diff
@@ -90,9 +90,8 @@ pb_bytes_array_t* Serializer::EncodeBytes(const std::vector<uint8_t>& bytes) {
   return ByteString(bytes).release();
 }
 
-std::vector<uint8_t> Serializer::DecodeBytes(const pb_bytes_array_t* bytes) {
-  if (bytes == nullptr) return {};
-  return std::vector<uint8_t>(bytes->bytes, bytes->bytes + bytes->size);
+ByteString Serializer::DecodeBytes(const pb_bytes_array_t* bytes) {
+  return ByteString(bytes);
```
Comment (rsgowman) [inline Firestore/core/src/firebase/firestore/remote/serializer.cc]: You've removed the `if (bytes==nullptr) return {};`. As it currently stands, this method will segfault if given a nullptr.

I don't think it was ever possible for nullptr to be passed. Even if the proto message is corrupted, I think you could get garbage values (incl 0s obviously) for the contents, but I think nanopb would always return *something*. I'm not 100% certain of that though.

I'd recommend either:
a) restoring the original behaviour.
b) hard_assert != nullptr. Either here or in the ByteString ctor (or both) and then adjusting the doc string in `serializer.h:118` to change this `If the input pointer is null, then this method will return an empty vector.` to `@pre bytes != nullptr` or similar.

File: Firestore/Source/Model/FSTFieldValue.mm [hunk 3]
```diff
@@ -748,7 +679,10 @@ - (id)value {
       HARD_FAIL("TODO(rsgowman): implement");
     case FieldValue::Type::String:
       return util::WrapNSString(self.internalValue.string_value());
-    case FieldValue::Type::Blob:
+    case FieldValue::Type::Blob: {
+      auto &blob = self.internalValue.blob_value();
```
Comment (rsgowman) [inline Firestore/Source/Model/FSTFieldValue.mm]: nit: The `blob` name doesn't seem useful to me, especially given the method that returns it is already named `blob_value()`. I'd either introduce the type (rather than auto) or just inline it with the return statement. Probably the latter. Optional.

File: Firestore/core/src/firebase/firestore/remote/serializer.cc (hunk 2 above)
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/remote/serializer.cc]: Ugh. `ByteString` previously sort of tolerated its bytes being null, though this was only the default constructed/moved-from state but not really intended for the empty case. Over time you've been pushing me to assert that the member `bytes_` is not `nullptr`. This particular case has me thinking that the approach is flawed.

In particular, if `ByteString` has to deal with instances of itself being valid that have no bytes, there seems little reason to force every caller to deal with this ambiguity.

Looking back at the docs, proto3 doesn't distinguish between empty and missing string/bytes values. To me, this means that our attempting to do so here is making a distinction that shouldn't really matter. Indeed I've added a test (`EncodesNullBlobs`) that shows that this behavior holds.

So, concretely, changes I've made in response to this feedback:
  * `ByteString` now explicitly avoids allocation in the empty case. `ByteString()` and `ByteString("")` are now indistinguishable.
  * Default, empty, and moved-from `ByteString`s all have the same representation.
  * Callers no longer have to handle null `pb_bytes_array_t*` values. This code is now correct as written.
  * I've stripped all the `HARD_ASSERT(bytes_ != nullptr)` out of the implementation.

Comment (rsgowman) [reply Firestore/core/src/firebase/firestore/remote/serializer.cc]: OTOH, consumers now have to handle null return values from ByteString.get() and .release(). (Though curiously not ToNSData() which will allocate a 0 length buffer.)

You could:
a) Change the default/etc ctor to always create a zero-length buffer. (Allowing input of nullptr and then convert to a zero-length buffer.) This would allow you to remove the special casing of nullptr as all methods could then assume a non-null `bytes_*`.
b) Change .get() and .release() to return non-null, even when bytes_ == nullptr. (get() could return a pointer to some static (constexpr?) value. release() would... I guess have to alloc. But I suppose that's what ToNSData() does now.)
c) Just make the callers deal with nullptrs. For consistency, consider altering ToNSData to also return nullptr when bytes_ == nullptr.

proto3 doesn't distinguish between empy/missing bytes on input, but when you retrieve the bytes, I believe it returns an empty byte array (rather than null) in both cases. If we're going to mimic it's behaviour, then this would argue for (a) or (b).

But at this point, we're bikesheding. What you've got is fine. Only requirement from me is to at least consider the above before hitting the submit button. :)

File: Firestore/core/src/firebase/firestore/nanopb/byte_string.h [hunk 4]
```diff
@@ -44,6 +44,10 @@ class ByteString : public util::Comparable<ByteString> {
  public:
   ByteString() = default;
 
+  /**
```
Comment (rsgowman) [inline Firestore/core/src/firebase/firestore/nanopb/byte_string.h]: One final question (unrelated to this line): We've got pretty strong ownership semantics around `pb_bytes_array_t* bytes_`. Would it make sense to formalize it a bit further by converting it to `std::unique_ptr<pb_bytes_array_t> bytes_`?

File: Firestore/core/src/firebase/firestore/nanopb/byte_string.h [hunk 5]
```diff
@@ -103,7 +107,7 @@ class ByteString : public util::Comparable<ByteString> {
   }
 
   bool empty() const {
-    return size() == 0;
+    return bytes_ == nullptr || bytes_->size == 0;
```
Comment (rsgowman) [inline Firestore/core/src/firebase/firestore/nanopb/byte_string.h]: line 122:
If you change things such that get() always returns non-nullptr, then you could alter the signature from `const pb_bytes_array_t* get() const` to `const pb_bytes_array_t& get() const`

File: Firestore/core/src/firebase/firestore/nanopb/byte_string.h [hunk 6]
```diff
@@ -122,15 +174,15 @@ class ByteString : public util::Comparable<ByteString> {
 
   util::ComparisonResult CompareTo(const ByteString& rhs) const;
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/nanopb/byte_string.h]: Perhaps comment that the comparison is lexicographical?

File: Firestore/core/src/firebase/firestore/nanopb/byte_string.h [hunk 7]
```diff
@@ -39,6 +44,11 @@ class ByteString : public util::Comparable<ByteString> {
  public:
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/nanopb/byte_string.h]: Can you please expand on the class comment? In particular, it might be worth that this class owns its memory,
has deep copy semantics, and under most circumstances creates a copy of the input given to its constructor. 

Additionally:
* from its interface, it seems like it's much more of a byte buffer than a string, because it doesn't provide any text manipulation. The only exception I noticed is that comparison is lexicographical;
* what's the intended use for this class? Why would somebody use it instead of manipulating Nanopb types directly? I'm mostly asking due to the presence of `release` function.

File: Firestore/core/test/firebase/firestore/remote/serializer_test.cc [hunk 8]
```diff
@@ -454,19 +454,45 @@ TEST_F(SerializerTest, EncodesTimestamps) {
...
+  for (const ByteString& blob_value : cases) {
+    FieldValue model = FieldValue::FromBlob(blob_value);
     ExpectRoundTrip(model, ValueProto(blob_value), FieldValue::Type::Blob);
   }
 }
 
+TEST_F(SerializerTest, EncodesNullBlobs) {
+  ByteString blob;
+  FieldValue model = FieldValue::FromBlob(blob);
+  ASSERT_EQ(blob.get(), nullptr);
```
Comment (var-const) [inline Firestore/core/test/firebase/firestore/remote/serializer_test.cc]: Optional: perhaps move this assertion one line up? As it stands, it seems to check that `blob` _became_ null after the creation of a field value.

File: Firestore/core/test/firebase/firestore/nanopb/byte_string_test.cc [hunk 9]
```diff
@@ -39,7 +39,26 @@ std::vector<uint8_t> ToVector(const std::string& str) {
 
 TEST(ByteStringTest, DefaultConstructor) {
   ByteString str;
-  EXPECT_EQ(nullptr, str.data());
+  EXPECT_EQ(str.get(), nullptr);
+
+  // Even though the backing bytes array is null, data()/size() should be
```
Comment (var-const) [inline Firestore/core/test/firebase/firestore/nanopb/byte_string_test.cc]: Nit: `size() should be non-null` probably needs to be rephrased.

File: Firestore/core/src/firebase/firestore/model/field_value.cc [hunk 10]
```diff
@@ -701,9 +688,8 @@ FieldValue FieldValue::FromString(std::string&& value) {
   return FieldValue(std::make_shared<StringValue>(std::move(value)));
 }
 
-FieldValue FieldValue::FromBlob(const uint8_t* source, size_t size) {
-  std::vector<uint8_t> copy(source, source + size);
-  return FieldValue(std::make_shared<BlobValue>(std::move(copy)));
+FieldValue FieldValue::FromBlob(ByteString blob) {
+  return FieldValue(std::make_shared<BlobValue>(std::move(blob)));
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/model/field_value.cc]: Can this be `std::make_shared<const BlobValue>`? Feel free to ignore if there's too much collateral compilation damage.

File: Firestore/core/src/firebase/firestore/local/query_data.h [hunk 11]
```diff
@@ -101,20 +102,20 @@ class QueryData {
...
  private:
-  const core::Query query_;
+  core::Query query_;
   model::TargetId target_id_;
   model::ListenSequenceNumber sequence_number_;
   QueryPurpose purpose_;
-  const model::SnapshotVersion snapshot_version_;
-  const std::vector<uint8_t> resume_token_;
+  model::SnapshotVersion snapshot_version_;
+  nanopb::ByteString resume_token_;
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/local/query_data.h]: I understand that `nanopb::ByteString` comes in handy here, but I'm not sure that:
* `nanopb` types should really propagate outside of `Serializer`;
* the malloc-created pointer to a nanopb-type is really the best representation for a resume token.

File: Firestore/core/src/firebase/firestore/nanopb/byte_string.cc [hunk 12]
```diff
@@ -85,19 +102,39 @@ ByteString::~ByteString() {
...
+  }
+}
+
 /* static */ ByteString ByteString::Take(pb_bytes_array_t* bytes) {
   return ByteString{bytes};
 }
 
+const uint8_t* ByteString::data() const {
+  static const uint8_t kEmpty[] = "";
+  return bytes_ ? bytes_->bytes : kEmpty;
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/nanopb/byte_string.cc]: Question: seems like `pb_bytes_array_t` is guaranteed to be `uint8_t*`?

File: Firestore/core/src/firebase/firestore/nanopb/byte_string.h [hunk 13]
```diff
@@ -71,34 +93,60 @@ class ByteString : public util::Comparable<ByteString> {
...
     return bytes_;
   }
 
   /**
-   * Returns the current byte array and assigns the backing byte array to
-   * nullptr, releasing the ownership of the array contents to the caller.
+   * Releases ownership of the backing byte array, and returns it to the caller.
+   * The backing byte array is set to null.
+   *
+   * This value may be null because nanopb (and protobuf generally) treat null
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/nanopb/byte_string.h]: Question: I might have missed something, but all the uses of `release` I remember from reading this PR were of the form:
```cpp
nanopb_type.bytes = ByteString{some_input).release(); // avoid double-free
```
In other words, `ByteString` was used as a way to create a deep copy of the pointer. This pattern seems like it's almost creating an object for a side effect, not to actually use it.

I don't know if there are other known uses of `release` in the future (or I missed something in this PR). If not, then it seems like providing a dedicated copy function would not only allow eliminating `release`, but perhaps allow changing the internal representation to `vector<uint8_t>`? On the other hand, it would make `Take` impossible -- I don't know how much of a performance hit that would be.

File: Firestore/core/src/firebase/firestore/nanopb/byte_string.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/nanopb/byte_string.h]: I wasn't sure I agreed when I started to try to follow this advice, but in the end I like it more.

Unfortunately, the class was so large because it was trying to make a bunch of disparate use cases fit together without changing the sources that much. Fixing this required fixing `ByteString` users to be more uniform.

For example, `ByteString` had constructors/conversions to deal with `std::vector` because that was how `ByteStringWriter` accumulated its contents. However, making `ByteStringWriter` directly accumulate in a `pb_bytes_array_t*` eliminates this need (and also saves a copy).

I've also made the conversions free functions since they can depend on the public interface.

File: Firestore/core/src/firebase/firestore/nanopb/byte_string.cc [hunk 14]
```diff
@@ -109,6 +146,10 @@ util::ComparisonResult ByteString::CompareTo(const ByteString& rhs) const {
   return util::Compare(absl::string_view{*this}, absl::string_view{rhs});
 }
 
+size_t ByteString::Hash() const {
+  return util::Hash(util::make_range(begin(), end()));
```
[the remark this answers is not in the document]
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/nanopb/byte_string.cc]: Without it there's an infinite recursion because `util::Hash` prefers `Hash()` members over hashing like a container.

File: Firestore/core/test/firebase/firestore/nanopb/byte_string_test.cc (hunk 9 above)
Comment (wilhuff) [reply Firestore/core/test/firebase/firestore/nanopb/byte_string_test.cc]: Removed `/size()`.

File: Firestore/core/src/firebase/firestore/nanopb/byte_string.cc (hunk 12 above)
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/nanopb/byte_string.cc]: `pb_bytes_array_t` is a struct with`pb_size_t` and `pb_byte_t[]` members. The latter is what we use as the address of the contents of the buffer.

File: Firestore/core/src/firebase/firestore/nanopb/byte_string.cc [hunk 15]
```diff
@@ -85,19 +102,39 @@ ByteString::~ByteString() {
...
 pb_bytes_array_t* ByteString::release() {
   pb_bytes_array_t* result = bytes_;
   bytes_ = nullptr;
   return result;
 }
 
 std::vector<uint8_t> ByteString::ToVector() const {
-  HARD_ASSERT(bytes_ != nullptr);
-  return std::vector<uint8_t>{bytes_->bytes, bytes_->bytes + bytes_->size};
+  return std::vector<uint8_t>{begin(), end()};
```
[the remark this answers is not in the document]
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/nanopb/byte_string.cc]: The difference is larger when `str` is an unnamed temporary.

```c++
auto str = MakeBytes();
std::vector<uint8_t> v{str.begin(), str.end()};

auto v = MakeBytes().ToVector();
```

I've resolved this by pulling `MakeVector` out into a free function along with the other conversions. This reduces clutter here while preserving this ease of use (which is done quite a bit in tests).

File: Firestore/core/src/firebase/firestore/nanopb/writer.h [hunk 16]
```diff
@@ -66,17 +63,43 @@ class Writer {
...
+  size_t remain() const {
+    return capacity_ - size();
+  }
+
+  size_t capacity() const {
+    return capacity_;
+  }
+
+  uint8_t* pos() {
+    return buffer_->bytes + buffer_->size;
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/nanopb/writer.h]: `size` checks for `buffer_` being null, should this function follow suit?

File: Firestore/core/src/firebase/firestore/nanopb/writer.cc [hunk 17]
```diff
@@ -72,20 +37,77 @@ void Writer::WriteNanopbMessage(const pb_field_t fields[],
   }
 }
 
-ByteStringWriter::ByteStringWriter() : Writer({}) {
-  stream_ = WrapContainer(&buffer_);
+static bool AppendToBytesArray(pb_ostream_t* stream,
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/nanopb/writer.cc]: Nit: prefer anonymous namespace (the `constexpr` constant above could be moved into it as well).

File: Firestore/core/src/firebase/firestore/nanopb/writer.cc [hunk 18]
```diff
@@ -72,20 +37,77 @@ void Writer::WriteNanopbMessage(const pb_field_t fields[],
...
+                               const pb_byte_t* buf,
+                               size_t count) {
+  auto writer = static_cast<ByteStringWriter*>(stream->state);
+  writer->Append(buf, count);
+  return true;
 }
 
-ByteString ByteStringWriter::ToByteString() const {
-  return ByteString(buffer_);
+ByteStringWriter::ByteStringWriter() : Writer() {
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/nanopb/writer.cc]: Nit: I don't think explicitly calling the default constructor of the base class is necessary.

File: Firestore/core/src/firebase/firestore/nanopb/writer.cc [hunk 19]
```diff
@@ -72,20 +37,77 @@ void Writer::WriteNanopbMessage(const pb_field_t fields[],
...
+  stream_.callback = AppendToBytesArray;
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/nanopb/writer.cc]: `AppendToBytesArray` can avoid `extern "C"` because we compile nanopb in C++ mode?

File: Firestore/core/src/firebase/firestore/nanopb/writer.cc [hunk 20]
```diff
@@ -72,20 +37,77 @@ void Writer::WriteNanopbMessage(const pb_field_t fields[],
...
-  return std::move(buffer_);
+ByteStringWriter::~ByteStringWriter() {
+  std::free(buffer_);
+}
+
+void ByteStringWriter::Append(const uint8_t* data, size_t size) {
+  if (size == 0) return;
+
+  Reserve(size);
+  uint8_t* pos = buffer_->bytes + buffer_->size;
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/nanopb/writer.cc]: Nit: this looks like it can be replaced with a call to `pos` member function.

File: Firestore/core/src/firebase/firestore/nanopb/writer.cc [hunk 21]
```diff
@@ -72,20 +37,77 @@ void Writer::WriteNanopbMessage(const pb_field_t fields[],
...
+  std::memcpy(pos, data, size);
+  buffer_->size += size;
+}
+
+void ByteStringWriter::Reserve(size_t size) {
+  size_t current_size = buffer_ ? buffer_->size : 0;
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/nanopb/writer.cc]: Why not call the member `size`?

File: Firestore/core/src/firebase/firestore/nanopb/writer.h [hunk 22]
```diff
@@ -66,17 +63,43 @@ class Writer {
...
+
+  ByteStringWriter(const ByteStringWriter&) = delete;
+  ByteStringWriter(ByteStringWriter&&) = delete;
+
+  ByteStringWriter& operator=(const ByteStringWriter&) = delete;
+  ByteStringWriter& operator=(ByteStringWriter&&) noexcept = delete;
 
-  nanopb::ByteString ToByteString() const;
+  void Append(const uint8_t* data, size_t size);
+  void Reserve(size_t size);
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/nanopb/writer.h]: `reserve` on `std::vector` takes the _total_ desired capacity, while this function takes a delta applied to the current size. Please call it out in a comment (or perhaps just change the behavior to mimic `vector`).

File: Firestore/core/src/firebase/firestore/nanopb/writer.cc [hunk 23]
```diff
@@ -72,20 +37,77 @@ void Writer::WriteNanopbMessage(const pb_field_t fields[],
...
+  capacity_ = desired;
+}
+
+void ByteStringWriter::SetSize(size_t size) {
+  HARD_ASSERT(size <= capacity_);  // Should have reserved.
+  buffer_->size = CheckedSize(size);
+}
+
+ByteString ByteStringWriter::Release() {
+  pb_bytes_array_t* pending = buffer_;
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/nanopb/writer.cc]: This should probably also zero out `capacity_`.

File: Firestore/core/test/firebase/firestore/nanopb/nanopb_testing.h [hunk 24]
```diff
@@ -35,21 +37,27 @@ namespace nanopb {
...
+  writer.Reserve(size);
+
+  bool ok =
+      protobuf_message.SerializeToArray(writer.pos(), static_cast<int>(size));
 
   // SerializeToArray can only fail if the buffer wasn't large enough.
   HARD_ASSERT(ok);
-  return nanopb::ByteString(buffer);
+
+  writer.SetSize(size);
```
Comment (var-const) [inline Firestore/core/test/firebase/firestore/nanopb/nanopb_testing.h]: This is a little error-prone, though I understand it's hard to avoid. One random idea is something like
```cpp
bool ok = writer.WriteDirectly(size, [&](uint8_t* pos, size_t size) {
  return protobuf_message.SerializeToArray(pos, static_cast<int>(size));
});

template <typename ReturnT, typename CallbackT>
auto ByteStringWriter::WriteDirectly(size_t size, CallbackT callback) -> decltype(callback()) {
  Reserve(size);
  auto result = callback(pos(), size);
  SetSize(size);
  return result;
}
```
Not necessarily worth the trouble, just throwing it out here. Maybe `pos` and `SetSize` (and even `Reserve`) could become private in that case.

File: Firestore/core/test/firebase/firestore/nanopb/writer_test.cc [hunk 25]
```diff
@@ -0,0 +1,57 @@
...
+    ASSERT_EQ(writer.capacity(), 8);
+  }
+
+  // Exceeding the doubled amount will resize to that
+  std::vector<uint8_t> large(20, 'a');
+  writer.Append(large.data(), large.size());
+  ASSERT_EQ(writer.size(), 28);
+  ASSERT_EQ(writer.capacity(), 28);
+}
+
```
Comment (var-const) [inline Firestore/core/test/firebase/firestore/nanopb/writer_test.cc]: Please add a test that exercises `Release`, showing that it leaves the writer empty and reassignable.

**Comment by var-const** [comment]
Very close to LGTM, but I wrote some feedback on the `ByteStringWriter`.

File: Firestore/core/test/firebase/firestore/nanopb/nanopb_testing.h (hunk 24 above)
Comment (var-const) [reply Firestore/core/test/firebase/firestore/nanopb/nanopb_testing.h]: Hmm, I've only noticed the usage in this file, where `SerializeToArray` appears to return a `bool`, have I missed something? However, I agree that at this point it's probably more trouble than it's worth; we can revisit later on if this functionality sees more usage.

Comment (wilhuff) [reply Firestore/core/test/firebase/firestore/nanopb/nanopb_testing.h]: SerializeToArray returns a bool but then we immediately assert success, because the only way we could have failed was to pass a buffer smaller than `message.ByteSizeLong()`. In my initial attempt of this, I just immediately made that assertion inside the lambda and returned nothing.

If we implement anything else that uses this API we should revisit, and how that interaction fits together might guide our redesign. For now I'm going to leave it.

File: Firestore/core/src/firebase/firestore/nanopb/writer.h (hunk 22 above)
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/nanopb/writer.h]: If there are more users of this API we can revisit. For now, matching `std::vector::reserve` is fine. I could change the name to `AddCapacity` if I wanted to change to the other model.
